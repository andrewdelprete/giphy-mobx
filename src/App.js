import React, { Component } from 'react';
import {observer, inject} from "mobx-react"
import {action} from "mobx"
import {getRandomGif} from './helpers/gifs'
import remove from 'lodash/remove'
import Button from './Button'
import Gif from './Gif'

@inject("gifs")
@observer
class App extends Component {
  @action.bound add() {
    getRandomGif().then(gif => this.props.gifs.push(gif))
  }

  @action.bound remove(gif) {
    remove(this.props.gifs, gif)
  }

  @action.bound removeLast() {
    this.props.gifs.pop()
  }

  render() {
    const { gifs } = this.props
    return (
      <div className="App">
        { gifs.length > 0 &&
          gifs.map((gif, i) => {
            return <Gif
              key={i}
              onClick={() => this.remove(gif) }
              style={{ backgroundImage: `url(${gif.images.fixed_width.url}` }}
            />
          }
        )}
        <div style={{ clear: 'both' }}>
          <Button primary style={{ margin: 0 }} onClick={this.add}>Add Gif</Button>
          <Button onClick={this.removeLast}>Remove Last</Button>
        </div>
      </div>
    );
  }
}

export default App;

