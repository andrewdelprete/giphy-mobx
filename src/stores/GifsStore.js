import { observable } from "mobx"
import {getRandomGif} from '../helpers/gifs'
import remove from 'lodash/remove'

class GifsStore {
  constructor() {
    this.gifs = observable([])
    this.add = this.add.bind(this)
    this.remove = this.remove.bind(this)
    this.removeLast = this.removeLast.bind(this)
  }

  add () {
    getRandomGif().then(gif => this.gifs.push(gif))
  }

  remove (gif) {
    remove(this.gifs, gif)
  }

  removeLast () {
    this.gifs.pop()
  }
}

export default new GifsStore()
