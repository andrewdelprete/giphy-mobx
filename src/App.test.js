import React from 'react';
import ReactDOM from 'react-dom';
import App from './App';
import {mount} from 'enzyme';
import toJson from 'enzyme-to-json';
import GifsStore from './stores/GifsStore'

function mockResponse (status, statusText, body) {
  return new window.Response(JSON.stringify(body), {
    status: status,
    statusText: statusText,
    headers: {
      'Content-type': 'application/json'
    }
  })
}

beforeEach(() => {
  const newGif = {
    images: {
      fixed_width: {
        url: 'new-it.gif'
      }
    }
  }

  // Mock fetch
  global.fetch = jest.fn().mockImplementation(() =>
     Promise.resolve(mockResponse(200, '', { data: [newGif] }))
  )
})

afterEach(() => {
  global.fetch.mockReset()
  GifsStore.gifs = []
})

it('adds a new gif when clicking add button', function(done) {
  let wrapper = mount(<App GifsStore={GifsStore} />)

  wrapper.find('#add').simulate('click')

  process.nextTick(() => {
    expect(GifsStore.gifs).toHaveLength(1)
    expect(toJson(wrapper)).toMatchSnapshot()
    done()
  })
})

it('should remove a gif when clicking it', function() {
  GifsStore.gifs = [{
    images: {
      fixed_width: {
        url: 'https://media1.giphy.com/media/we1KGq2yvN65a/200w.gif'
      }
    }
  }]

  let wrapper = mount(<App GifsStore={GifsStore} />)

  wrapper.find('a').simulate('click')

  expect(GifsStore.gifs).toHaveLength(0)
})

it('should remove last gif when clicking it', function() {
  GifsStore.gifs = [
    {
      images: {
        fixed_width: {
          url: 'new-gif.gif'
        }
      }
    },
    {
      images: {
        fixed_width: {
          url: 'new-gif-2.gif'
        }
      }
    }
  ]

  let wrapper = mount(<App GifsStore={GifsStore} />)

  expect(GifsStore.gifs).toHaveLength(2)
  wrapper.find('#removeLast').simulate('click')
  expect(GifsStore.gifs).toHaveLength(1)
})

it('should render no gifs', () => {
  let wrapper = mount(<App GifsStore={GifsStore} />)
  expect(GifsStore.gifs).toHaveLength(0)
  expect(toJson(wrapper)).toMatchSnapshot()
});

it('should render one gif', () => {
  GifsStore.gifs = [{
    images: {
      fixed_width: {
        url: 'https://media1.giphy.com/media/we1KGq2yvN65a/200w.gif'
      }
    }
  }]

  var wrapper = mount(<App GifsStore={GifsStore} />)

  expect(toJson(wrapper)).toMatchSnapshot()
})