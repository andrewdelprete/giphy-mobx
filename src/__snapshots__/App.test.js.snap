exports[`test adds a new gif when clicking add button 1`] = `
<Component
  GifsStore={
    GifsStore {
      "add": [Function],
      "gifs": Array [
        Object {
          "images": Object {
            "fixed_width": Object {
              "url": "new-it.gif",
            },
          },
        },
      ],
      "remove": [Function],
      "removeLast": [Function],
    }
  }>
  <div
    className="App">
    <div>
      <styled.a
        onClick={[Function]}
        style={
          Object {
            "backgroundImage": "url(new-it.gif)",
          }
        }>
        <a
          className="juSYaJ"
          onClick={[Function]}
          style={
            Object {
              "backgroundImage": "url(new-it.gif)",
            }
          } />
      </styled.a>
    </div>
    <div
      style={
        Object {
          "clear": "both",
        }
      }>
      <styled.button
        id="add"
        onClick={[Function]}
        primary={true}
        style={
          Object {
            "margin": 0,
          }
        }>
        <button
          className="kioqWX"
          id="add"
          onClick={[Function]}
          style={
            Object {
              "margin": 0,
            }
          }>
          Add Gif
        </button>
      </styled.button>
      <styled.button
        id="removeLast"
        onClick={[Function]}>
        <button
          className="iOiCLo"
          id="removeLast"
          onClick={[Function]}>
          Remove Last
        </button>
      </styled.button>
    </div>
  </div>
</Component>
`;

exports[`test should render no gifs 1`] = `
<Component
  GifsStore={
    GifsStore {
      "add": [Function],
      "gifs": Array [],
      "remove": [Function],
      "removeLast": [Function],
    }
  }>
  <div
    className="App">
    <div />
    <div
      style={
        Object {
          "clear": "both",
        }
      }>
      <styled.button
        id="add"
        onClick={[Function]}
        primary={true}
        style={
          Object {
            "margin": 0,
          }
        }>
        <button
          className="kioqWX"
          id="add"
          onClick={[Function]}
          style={
            Object {
              "margin": 0,
            }
          }>
          Add Gif
        </button>
      </styled.button>
      <styled.button
        id="removeLast"
        onClick={[Function]}>
        <button
          className="iOiCLo"
          id="removeLast"
          onClick={[Function]}>
          Remove Last
        </button>
      </styled.button>
    </div>
  </div>
</Component>
`;

exports[`test should render one gif 1`] = `
<Component
  GifsStore={
    GifsStore {
      "add": [Function],
      "gifs": Array [
        Object {
          "images": Object {
            "fixed_width": Object {
              "url": "https://media1.giphy.com/media/we1KGq2yvN65a/200w.gif",
            },
          },
        },
      ],
      "remove": [Function],
      "removeLast": [Function],
    }
  }>
  <div
    className="App">
    <div>
      <styled.a
        onClick={[Function]}
        style={
          Object {
            "backgroundImage": "url(https://media1.giphy.com/media/we1KGq2yvN65a/200w.gif)",
          }
        }>
        <a
          className="juSYaJ"
          onClick={[Function]}
          style={
            Object {
              "backgroundImage": "url(https://media1.giphy.com/media/we1KGq2yvN65a/200w.gif)",
            }
          } />
      </styled.a>
    </div>
    <div
      style={
        Object {
          "clear": "both",
        }
      }>
      <styled.button
        id="add"
        onClick={[Function]}
        primary={true}
        style={
          Object {
            "margin": 0,
          }
        }>
        <button
          className="kioqWX"
          id="add"
          onClick={[Function]}
          style={
            Object {
              "margin": 0,
            }
          }>
          Add Gif
        </button>
      </styled.button>
      <styled.button
        id="removeLast"
        onClick={[Function]}>
        <button
          className="iOiCLo"
          id="removeLast"
          onClick={[Function]}>
          Remove Last
        </button>
      </styled.button>
    </div>
  </div>
</Component>
`;
